### PR TITLE
Vectorize

### PR DESCRIPTION
**Pull request by jeanconn**
Redo vectorized approach.

### CODE REVIEW

**Comment by jeanconn** [comment]
Code needs comments.

File: Quaternion/Quaternion.py [hunk 1]
```diff
@@ -100,10 +99,24 @@ class Quat(object):
...
+     * an N x 4 element array ( N by x,y,z,w quat form)
+     * an N x 3 element array ( N by ra, dec, roll in degrees,
+       if N == 3, the optional 'intype = 'equatorial' may be used to differentiate,
+       this from a transform matrix)
+     * an N x 3x3
+
+   :param intype: optional type to describe input attitude
+
+   ``intype`` may be:
+     * transform
```
Comment (jeanconn) [inline Quaternion/Quaternion.py]: This should probably be quoted.

File: Quaternion/Quaternion.py [hunk 2]
```diff
@@ -100,10 +99,24 @@ class Quat(object):
...
+     * equatorial
+     * quaternion
+
 
     """
 
-    def __init__(self, attitude):
+    def __init__(self, attitude, intype=None):
```
Comment (jeanconn) [inline Quaternion/Quaternion.py]: The idea of "intype" is really to let us easily keep the default behavior that a 3x3 is treated as a transform and not a weird three-row ra,dec,roll initialization.  We could probably also check for a transform matrix vs ra,dec,roll entries by value, but the initialization option seemed a safe convention at the time.

File: Quaternion/Quaternion.py [hunk 3]
```diff
@@ -113,34 +126,44 @@ def __init__(self, attitude):
...
         """
         Set the value of the 4 element quaternion vector
+        May be 4 element list or array or N x 4 element array.
 
         :param q: list or array of normalized quaternion elements
         """
         q = np.array(q)
-        if abs(np.sum(q**2) - 1.0) > 1e-6:
+        if q.ndim == 1:
+            q = q[np.newaxis]
```
Comment (jeanconn) [inline Quaternion/Quaternion.py]: Throughout, this "vectorized" approach tries to take the single-attitude case, convert it to a 1xN with newaxis for most of the internal operations, and then return just the single-attitude again again.  This was both for convenience and backwards compatibility, but we could go other ways.

File: Quaternion/Quaternion.py [hunk 4]
```diff
@@ -151,15 +174,17 @@ def _get_q(self):
...
         return self._q
 
     # use property to make this get/set automatic
     q = property(_get_q, _set_q)
 
-    def __repr__(self):
-        q = self.q
-        return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
-                .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+    #def __repr__(self):
```
Comment (jeanconn) [inline Quaternion/Quaternion.py]: Not sure about the best way to define the repr for N entries.

**Comment by jeanconn** [comment]
For a bunch of current work, it would probably be helpful to have the one, vectorized Quaternion approach instead of a mix of current Quaternion and mica.quaternion whenever vectors are needed.  I'm not sure if there are any showstoppers in this implementation.

**Comment by jeanconn** [comment]
Needs updated tests.

**Comment by jeanconn** [comment]
Move to explicit keywords for the input types.

**Comment by jeanconn** [comment]
Does this also need rebase?

**Comment by javierggt** [comment]
or merge. What I did in my local copy for now is to merge into master. Otherwise the test does not work.

**Comment by taldcroft** [comment]
Our workflow (inherited from astropy) is always rebase to get rid of conflicts, never merge master.  I'm not precisely sure of the reason, but in astropy anyway whenever contributors accidentally do a merge it tends to be a disaster and they get like 50 unrelated commits showing up in their PR.

**Comment by javierggt** [comment]
but one should not rebase a branch that was already pushed. I have done merges to branches without problem.

In any case, I did the merge after pushing, so it should not be in github.

Right now gitk kills my OSX session, so I can't inspect the index before pushing, which I normally do. Hopefully I did not break anything.

**Comment by javierggt** [comment]
by the way, I understand about not merging master, I am not advocating it. I suppose cherry-picking from master is the right way. If we rebase on a branch that was pushed already, shouldn't one then  create a whole new branch?

**Comment by jeanconn** [comment]
Given that is is just us with this branch and we're likely to have one developer at a time working in the branch, I'm good with force pushing the branch after rebase.

**Comment by taldcroft** [comment]
Agreed with @jeanconn.  Even in the wild world of astropy we rebase and force push branches all the time.  Reviewers can just delete their local version and re-fetch as needed. 

The only hard rule is never force push to master or a release branch.

**Comment by javierggt** [comment]
ok. Force-pushed.

When making the changes for arbitrary shapes, there was some juggling with axes, because the transpose function does not do what we want. I used np.moveaxis. I think I can go over the code once again to see if that can be written better.

**Comment by taldcroft** [comment]
Yeah, the axis juggling was why I had thought about numba + simple loops on the outside.  But if you are able to get it right with axes then that's great.

File: Quaternion/Quaternion.py [hunk 5]
```diff
@@ -275,38 +326,35 @@ def _quat2equatorial(self):
...
-        if (roll < 0):
-            roll += 360
-
-        return np.array([ra, dec, roll])
+        ra = np.degrees(np.arctan2(xb, xa))
+        dec = np.degrees(np.arctan2(xn, np.sqrt(one_minus_xn2)))
+        roll = np.degrees(np.arctan2(yn, zn))
+        ra[ra < 0] = ra[ra < 0] + 360
+        roll[roll < 0] = roll[roll < 0] + 360
+        return np.moveaxis(np.array([ra, dec, roll]), 0, -1)
```
Comment (taldcroft) [inline Quaternion/Quaternion.py]: Add a comment explaining what's happening here.  Not totally obvious to me.

File: Quaternion/Quaternion.py [hunk 6]
```diff
@@ -152,15 +193,19 @@ def _get_q(self):
...
     q = property(_get_q, _set_q)
 
     def __repr__(self):
         q = self.q
-        return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
-                .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+        if q.ndim == 1:
+            return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
+                    .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+        return ('{}({})'.format(self.__class__.__name__, repr(q)))
```
[the remark this answers is not in the document]
Comment (taldcroft) [reply Quaternion/Quaternion.py]: Per discussion in our meeting this comment is rescinded.

File: Quaternion/Quaternion.py [hunk 7]
```diff
@@ -275,38 +326,35 @@ def _quat2equatorial(self):
...
         Determine Right Ascension, Declination, and Roll for the object quaternion
 
-        :returns: RA, Dec, Roll
+        :returns: N x (RA, Dec, Roll)
         :rtype: numpy array [ra,dec,roll]
         """
 
-        q = self.q
-        q2 = self.q**2
+        q = np.atleast_2d(self.q)
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: Yes, but note that nothing is being allocated, because atleast_2d returns a view.

It really depends on what you want. You could prefer to encapsulate all uses of self._q inside _set_q and _get_q, and use self.q everywhere else (the proper way of lazy initialization, IMO). Or you could want to optimize by removing the intermediate calls.

The current implementation has three properties (q, equatorial, transform), with corresponding lazy data members (_q, _equatorial, _transform). The lazy data members are only used in the constructor and the _get/set_ methods, so the get/set methods are tightly bound.

Using self._q here would break the pattern. This would be the one case where a lazy data member is called out of the get/setters. This function (_quat2equatorial) is only called from _get_equatorial, right after setting self._q anyway, so it happens to work. I would argue that it is this statement inside _get_equatorial that is redundant:
self._q = self._transform2quat()

File: Quaternion/Quaternion.py (hunk 5 above)
Comment (javierggt) [reply Quaternion/Quaternion.py]: I will add this:
```
        # all negative angles are incremented by 360,
        # the output is in the (0,360) interval instead of in (-180, 180)
```

File: Quaternion/Quaternion.py [hunk 8]
```diff
@@ -170,7 +215,7 @@ def _set_equatorial(self, equatorial):
            :param equatorial: list or array [ RA, Dec, Roll] in degrees
 
         """
-        self._equatorial = np.array(equatorial)
+        self._equatorial = np.atleast_2d(np.array(equatorial))
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: right

File: Quaternion/Quaternion.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (taldcroft) [reply Quaternion/Quaternion.py]: Fair enough.  What caught my attention is the undo-ing of `atleast_2d` in `_get_q` (in the scalar case) and then immediately re-doing it here.  But I agree what you have is more robust and consistent.

File: Quaternion/Quaternion.py (hunk 5 above)
Comment (taldcroft) [reply Quaternion/Quaternion.py]: I was referring to the `moveaxis`.

Comment (javierggt) [reply Quaternion/Quaternion.py]: ah! right. Not trivial.

It is a transpose. The resulting ra, dec roll have shape (N,) (where N is an arbitrary tuple), then they are stacked so the result is (3,N), moving the axis produces (N,3).

A transpose of (3,2,5) would produce (5,2,3) which is not what we want: (2,5,3).

Perhaps it is better to just create the result array first, and then fill it at the right position. No need to transpose then.

File: Quaternion/tests/test_all.py [hunk 9]
```diff
@@ -1,12 +1,60 @@
...
+import pytest
+
 from .. import Quat
 
 ra = 10.
 dec = 20.
 roll = 30.
 q0 = Quat([ra, dec, roll])
 
+equatorial_23 = np.array([[[ 10,  20,  30],
```
Comment (taldcroft) [inline Quaternion/tests/test_all.py]: For validation that the *vectorized operations* are working, we can stipulate that the scalar code works by seeing that the existing tests pass.  At that point the actual values don't matter so much as confirming that doing things using the new interface with non-scalar inputs gives the same answer as doing through the old scalar interface (on a per-element basis).  So something like (untested!):
```
equatorial_23 = np.linspace(1, 50, 2 * 3 * 3).reshape(2, 3, 3)
q_23 = np.empty(shape=(2, 3, 4))
transform_23 = np.empty(shape=(2, 3, 3, 3))
for ii in range(2):
    for jj in range(3):
        q = Quat(equatorial_23[ii, jj])
        q_23[ii, jj] = q.q
        transform_23[ii, jj] = q.transform
```
This should let you compare the vector interface to the scalar version to machine precision instead of `allclose` (which by default is only 6 digits).  (The plan is that the exact same math gets done at the end of the day).  Since these values are well-behaved you should be able to do round-trip tests and get answers that match to within pretty tight tolerances (one part in 1e-10 or better).

The idea of including the bad corner cases in the vector testing is good, but maybe just do that as a separate one-off.  Could be more minimal, just `[[10, -90, 30], [10, 90, -30]]`.

**Comment by javierggt** [comment]
I added a _shape data member (instead of _is_scalar). I figure it is more explicit. Then I replaced the if+squeeze statements by a reshape operation. If anyone has strong feelings about it, I can introduce the squeeze statements instead.

File: Quaternion/Quaternion.py [hunk 10]
```diff
@@ -101,47 +100,89 @@ class Quat(object):
       * a 4 element array (expects x,y,z,w quat form)
       * a 3 element array (expects ra,dec,roll in degrees)
       * a 3x3 transform/rotation matrix
+    N x those types :
+     * an N x 4 element array ( N by x,y,z,w quat form)
+     * an N x 3 element array ( N by ra, dec, roll in degrees,
+       if N == 3, the optional 'intype = 'equatorial' may be used to differentiate,
+       this from a transform matrix)
+     * an N x 3x3
+
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: done

File: Quaternion/Quaternion.py [hunk 11]
```diff
@@ -101,47 +100,89 @@ class Quat(object):
...
 
     """
 
-    def __init__(self, attitude):
+    def __init__(self, attitude=None, transform=None, q=None, equatorial=None):
+        npar = (int(attitude is not None) + int(transform is not None) +
+                int(q is not None) + int(equatorial is not None))
+        if npar != 1:
+            raise Exception('One and only one of attitude, transform, quaternion, equatorial must be given ({})'.format(npar))
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: done

File: Quaternion/Quaternion.py [hunk 12]
```diff
@@ -152,15 +193,19 @@ def _get_q(self):
                 self._q = self._equatorial2quat()
             elif self._T is not None:
                 self._q = self._transform2quat()
+        if self._q.shape[0] == 1:
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: as mentioned below. I added self._shape instead of _is_scalar.

File: Quaternion/Quaternion.py [hunk 13]
```diff
@@ -152,15 +193,19 @@ def _get_q(self):
...
         return self._q
 
     # use property to make this get/set automatic
     q = property(_get_q, _set_q)
 
     def __repr__(self):
         q = self.q
-        return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
-                .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+        if q.ndim == 1:
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/Quaternion.py]: this q variable already has the right shape. So q.ndim == 1 if and only if self._is_scalar.

File: Quaternion/Quaternion.py (hunk 5 above)
Comment (taldcroft) [reply Quaternion/Quaternion.py]: With the comment it is clear enough, but true that creating / filling in the output would be easier to read and not *need* a comment.  Either way is fine.

File: Quaternion/Quaternion.py [hunk 14]
```diff
@@ -479,6 +550,9 @@ def __mul__(self, quat2):
 
           q = Quat(numpy.dot(q1.transform, q2.transform))
```
Comment (taldcroft) [inline Quaternion/Quaternion.py]: Might as well update this to the modern nicer version `q = Quat(q1.transform @ q2.transform)` (but your caveat definitely applies).

**Comment by taldcroft** [comment]
For performance in the scalar case, I think the only really critical case is annie.  So one could just do a 10 ksec sim with this branch and with release and see if it matters.

**Comment by taldcroft** [comment]
@jeanconn - could you take a quick look at annie performance?

File: Quaternion/Quaternion.py (hunk 14 above)
Comment (javierggt) [reply Quaternion/Quaternion.py]: I think numpy.dot is not correct here. It should be as you say or numpy.matmul.

Also, docstrings should use the new API. So this should be q = Quat(transform=q1.transform @ q2.transform)

**Comment by jeanconn** [comment]
Regarding annie performance, sure happy to take an independent look, though I note the "classic use case" @javierggt has also calls an annie sim (my take-away being he is set up to do that annie benching too if I'm not fast enough for this thread).

File: Quaternion/Quaternion.py (hunk 14 above)
Comment (taldcroft) [reply Quaternion/Quaternion.py]: Agreed, dot only works in the scalar case where the transforms are 3x3.

**Comment by taldcroft** [comment]
@jeanconn - ah, then nevermind, @javierggt can just do it.

**Comment by javierggt** [comment]
so... stupid question...

I got the code that uses annie. I suppose that what you would like to time is this part:
```
sc = Spacecraft()
sc.track_stars_setup(**annie_kwargs)
sc.run()
```
Right?

**Comment by jeanconn** [comment]
That makes sense to me.  For that somewhat "canonical" example, the loop I had at the bottom that calculates delta quaternions

```
est_delta_pitchs = []
est_delta_yaws = []
times = []
for att_record in sc.telem.att_records:
    if att_record.att_est is not None:
        # For arbitrary lists of quats, we'll often just have
        # the 4 vector, even if the annie example already
        # has Quat objects for est, true, and cmd.  Here, I've just used the
        # ".q" of these Quat objects to reinitialize them.  I don't know if
        # it makes any difference for your testing
        dq = Quat(att_record.att_est.q).dq(att_record.att_true.q)
        est_delta_pitchs.append(dq.pitch * 3600)
        est_delta_yaws.append(dq.yaw * 3600)
        times.append(att_record.time)
```

is the kind of *thing* we do with the annie data, but in previous use that happened outside the simulation and as such the loop doesn't need to be timed; especially if we'd have a new recommended new way of doing it with the new code.

For benching the annie sims, there might be value in timing the 1000s simulation I had in that example and timing/benching a longer one (20ks?) just to check that nothing is weird with how it scales.

**Comment by javierggt** [comment]
right, but that loop I am already benchmarking it separately. I ended up splitting the statement
```
dq = Quat(att_record.att_est.q).dq(att_record.att_true.q)```

into
```
        q = Q(quaternions_true[i])
        dq = q.dq(quaternions_est[i])
```
so one can see each step

**Comment by jeanconn** [comment]
Yep.  That's what I meant by that loop doesn't need to be timed and we'd have a new way of doing it :+1:

**Comment by javierggt** [comment]
What way would be best to share a notebook? I can put it in google drive.

The bottomline is sc.run() is the same for all versions. There might be differences in the loop that comes afterwards

Each method (non-vectorized, running over quaternions one by one) timed:
https://drive.google.com/open?id=1IS5KJ1iwB_cTkJZfa5y92GoF03g75OwN

Annie timed:
https://drive.google.com/open?id=13_6ifBY4DPRaOyfYyFQ-bSRzv3kDCElR

File: Quaternion/Quaternion.py [hunk 15]
```diff
@@ -7,7 +7,7 @@
    - class methods to multiply and divide quaternions
 
 :Copyright: Smithsonian Astrophysical Observatory (2010)
-:Author: Jean Connelly (jconnelly@cfa.harvard.edu)
+:Authors: Tom Aldcroft (aldcroft@cfa.harvard.edu), Jean Connelly (jconnelly@cfa.harvard.edu)
```
Comment (jeanconn) [inline Quaternion/Quaternion.py]: @javierggt Now needs to be in here too!

**Comment by jeanconn** [comment]
For notebooks we don't have a one-size-fits-all when it comes to sharing.  If you think they will have value the next time we work in Quaternion, I think they could live in the repo.  If they are one-offs, google drive is fine though if they can be public, often we'll copy to a web area and serve with the web server (and then share links via nbviewer).  Sometimes we'll check in a notebook into the separate skanb project, though that method has not taken off.

File: Quaternion/Quaternion.py (hunk 15 above)
Comment (javierggt) [reply Quaternion/Quaternion.py]: done, but the line is now too long. I am not used to this docstring format, so I do not know how off the top of my head how to make it in multiple lines.

**Comment by javierggt** [comment]
ok, I made most changes.

What is left is:
- what @taldcroft mentioned above about validation of vectorized operations.
- something I do not understand yet (it is commented in the test).

File: Quaternion/tests/test_all.py (hunk 9 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: To do these "round-trip" check, I wrote a little generator helper function (maybe something exists already?) so I write something like

```
q1_inv = q1.inv()
for i in indices(q1._shape):
    # check that Quat(q).inv().q[i] == Quat(q[i]).inv().q
    assert np.all(q1_inv.q[i] == Quat(q=q1.q[i]).inv().q)
```

and 

```
q = Quat(equatorial=equatorial_23)
for i in indices(q._shape):
    # check that Quat(equatorial).q[i] == Quat(equatorial[i]).q
    assert np.all(q.q[i] == Quat(equatorial_23[i]).q)
```

Note that in this example I used the (private) _shape data member, so I am thinking of making it public. I do not use _shape in the version I am about to check in the repository.

File: Quaternion/Quaternion.py (hunk 15 above)
Comment (taldcroft) [reply Quaternion/Quaternion.py]: Looks like you can do this:
```
:Authors: - Me
          - Myself
          - I
```

File: Quaternion/tests/test_all.py [hunk 16]
```diff
@@ -30,6 +169,59 @@ def test_from_transform():
...
+    assert q.q.shape == (1, 1, 4)
+
+
+def test_eq_from_transform():
+    # this raises 'Unexpected negative norm' exception due to roundoff in copy/paste above
+    # q = Quat(transform=transform_23)
+    # assert q.equatorial.shape == (2, 3, 3)
+    # assert np.allclose(q.equatorial, equatorial_23)
+
+    # this one fails (quaternion -> equatorial -> quaternion is not an identity)
```
Comment (taldcroft) [inline Quaternion/tests/test_all.py]: This is now tracked with #10 and you can remove these comments and not worry about it for this PR.

File: Quaternion/Quaternion.py [hunk 17]
```diff
@@ -156,21 +158,21 @@ def __init__(self, attitude=None, transform=None, q=None, equatorial=None):
         # checking correct shapes
         if q is not None:
             q = np.atleast_1d(q)
-            self._shape = q.shape[:-1]
+            self.shape = q.shape[:-1]
```
Comment (taldcroft) [inline Quaternion/Quaternion.py]: Making this public is a great idea!  Maybe it should be a read-only property in case someone thinks they can reshape by setting `shape` the way you can in numpy.

Comment (javierggt) [reply Quaternion/Quaternion.py]: right, it should be read-only.

After I started this, I began wondering how much "array" interface this class should have, although what we have now is the minimal approach and should limit the scope of this merge.

Examples of array interface can include:
- the ability to multiply/divide/dx scalar quaternions by multi-quaternions (not allowed right now)
- the ability to iterate over quaternions in a multi-quaternion.
- broadcasting to multiply/divide/dx arbitrarily-sized multi-quaternions.

This will not be done for this merge, but you can think about it.

File: Quaternion/tests/test_all.py (hunk 9 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: ok, in commit 48c466dbfe5cd I removed these extra arrays.

Let it be noted that I do not like it. In my opinion, it decreases readability, it does not improve testing (the tests already were checking with full precision) and there is non-trivial code running out of tests.

**Comment by javierggt** [comment]
I do not intend to make more changes in this branch. All remaining issues are reported in #11, #12, #13 and #14.

File: Quaternion/tests/test_all.py [hunk 18]
```diff
@@ -16,6 +89,62 @@ def test_from_eq():
...
     assert np.allclose(q.roll0, 30)
     assert np.allclose(q.ra0, 10)
+    assert q.pitch == -q.dec
+    assert q.yaw == q.ra0
+
+    q1 = Quat(equatorial=[10, 90, 30])
+    q2 = Quat(q=q1.q)
+
+
+def test_from_eq_vectorized():
```
Comment (taldcroft) [inline Quaternion/tests/test_all.py]: The unrolled tests below are excellent.  Can you also add the tests like below, that are the most simple way for my brain to be sure that everything works as expected.  You might need to remove the dec=90 values from the input, not sure.  These are at least partially redundant, but not entirely (I think).
```
atol = 1e-12  # Hopefully this works?

# Compare vectorized computations for all possible input/output combos
# with the same for the scalar calculation.
q = Quat(equatorial=equatorial23)
assert np.all(q.q == q23)
assert np.all(q.equatorial == equatorial23)
assert np.all(q.transform == transform23)

q = Quat(q=q23)
assert np.all(q.q == q23)
assert np.allclose(q.equatorial, equatorial23, rtol=0, atol=atol)
assert np.allclose(q.transform, transform23, rtol=0, atol=atol)

q = Quat(transform=transform23)
assert np.allclose(q.q, q23, rtol=0, atol=atol)
assert np.allclose(q.equatorial, equatorial23, rtol=0, atol=atol)
assert np.all(q.transform == transform23)
```

File: Quaternion/tests/test_all.py [hunk 19]
```diff
@@ -101,3 +317,36 @@ def test_div_mult():
...
+    q1 = Quat(q=q_23[:1, :2])  # (shape (2,1)
+    q2 = Quat(q=q_23[1:, 1:])  # (shape (2,1)
+    assert q1.q.shape == q2.q.shape
+    q12 = q1*q2
+    assert q12.q.shape == q1.q.shape
+
+
+def test_normalize():
+    a = [[[1., 0., 0., 1.]]]
+    _ = Quat(q=normalize(a))
```
Comment (taldcroft) [inline Quaternion/tests/test_all.py]: Need to check the result.

File: Quaternion/tests/test_all.py (hunk 9 above)
Comment (taldcroft) [reply Quaternion/tests/test_all.py]: Noted.  

For me as the code reviewer reading the original code, the big glop of numbers was meaningless since I have no clue how they were generated or what they mean, and it was difficult to discern by eye the dimensionality of the arrays.  If you included a comment that showed the exact code to generate those values (and e.g. said that you ran it with the  current *release* version of Quaternion) then I might let it pass.

With the programmatic version the I can see instantly what's happening, in particular that the scalar version of Quaternion is being used to generate a multi-dimensional (list of lists) of reference values.

Just a difference in perspective between the person writing vs. reading the code.  :smile:

File: Quaternion/tests/test_all.py [hunk 20]
```diff
@@ -30,6 +159,59 @@ def test_from_transform():
     assert np.allclose(q.roll0, 30)
     assert np.allclose(q.ra0, 10)
 
+    t = [[9.25416578e-01,  -3.18795778e-01,  -2.04874129e-01],
+         [1.63175911e-01,   8.23172945e-01,  -5.43838142e-01],
+         [3.42020143e-01,   4.69846310e-01,   8.13797681e-01]]
+    _ = Quat(t)
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/tests/test_all.py]: No idea. I am removing it. A few lines above there is a test for this interface.

File: Quaternion/tests/test_all.py (hunk 18 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: This test fails because of the points I chose, which are gimbal locks.

Then I changed the points to some pseudo-random values and it also fails because q.equatorial can return roll=-30 or roll=330 depending on I don't know what. Angles are not supposed to be normalized? If not, I can check that they are the same module 360 or something like that.

File: Quaternion/tests/test_all.py [hunk 21]
```diff
@@ -30,6 +169,59 @@ def test_from_transform():
...
+    # assert np.allclose(q.equatorial, equatorial_23)
+
+    # this one fails (quaternion -> equatorial -> quaternion is not an identity)
+    # q = [0.24184476, -0.66446302, 0.24184476, 0.66446302]
+    # assert np.allclose(q, Quat(equatorial=Quat(q=q).equatorial).q)
+
+    t = np.zeros((4, 5, 3, 3))
+    t[:] = q0.transform[np.newaxis][np.newaxis]
+    q = Quat(transform=t)
+    print('roll', q.roll0)
```
[the remark this answers is not in the document]
Comment (javierggt) [reply Quaternion/tests/test_all.py]: removed

File: Quaternion/tests/test_all.py (hunk 16 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: removed

File: Quaternion/tests/test_all.py (hunk 18 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: and this one fails because some rows in q_23 have negative scalar component, which gets flipped in setter:
```
assert np.all(q.q == q23)
```

Comment (javierggt) [reply Quaternion/tests/test_all.py]: so... I added a helper function to deal with roll values out of the (0, 360) interval. Perhaps this should be fixed in the class.

I am also checking that q gets flipped when needed.

File: Quaternion/tests/test_all.py (hunk 19 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: fixed

**Comment by taldcroft** [comment]
Sadly, developer is not allowed to review.  :smile:  I think there is a way to enforce this in GitHub, but we just do this procedurally.  You can instead apply the "Ready for final review" label when you think it is finally done.

**Comment by taldcroft** [comment]
Will review on Monday (trying to make time for "science" today).

File: Quaternion/tests/test_all.py (hunk 18 above)
Comment (taldcroft) [reply Quaternion/tests/test_all.py]: I added this test and it is passing.  Look OK @javierggt ?

File: Quaternion/tests/test_all.py [hunk 22]
```diff
@@ -309,6 +309,35 @@ def test_dq_vectorized():
...
+    assert np.all(q.q == q_23)
+    assert np.allclose(q.equatorial, eq_23, rtol=0, atol=atol)
+    assert np.allclose(q.transform, transform_23, rtol=0, atol=atol)
+
+    q = Quat(transform=transform_23)
+    assert np.allclose(q.q, q_23, rtol=0, atol=atol)
+    assert np.allclose(q.equatorial, eq_23, rtol=0, atol=atol)
+    assert np.all(q.transform == transform_23)
+
+
```
Comment (javierggt) [inline Quaternion/tests/test_all.py]: all these test were already there:
- The first three assert are in [line 131 and on](https://github.com/sot/Quaternion/blob/c2c7bfc690407cbfac9016b8344a0c8fa348d026/Quaternion/tests/test_all.py#L131)
- the next three (including the angle normalization) are in [line 224 and on](https://github.com/sot/Quaternion/blob/c2c7bfc690407cbfac9016b8344a0c8fa348d026/Quaternion/tests/test_all.py#L224)
- the last three are in [line 195 and on](https://github.com/sot/Quaternion/blob/c2c7bfc690407cbfac9016b8344a0c8fa348d026/Quaternion/tests/test_all.py#L195)

Having them together means the angle normalization is done only once.

File: Quaternion/tests/test_all.py (hunk 18 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: reply to this is below.

File: Quaternion/tests/test_all.py (hunk 22 above)
Comment (taldcroft) [reply Quaternion/tests/test_all.py]: Ah I see, excellent.  But it doesn't hurt to have redundant tests, and these are organized in a way that works for my brain.  As long as you agree these tests are valid then I'd like to keep them in.

**Comment by taldcroft** [comment]
@javierggt - is the flipping thing something you can address in a post-merge fix?  From my perspective this looks good to merge now.

File: Quaternion/tests/test_all.py (hunk 22 above)
Comment (javierggt) [reply Quaternion/tests/test_all.py]: I'm ok with that.

**Comment by javierggt** [comment]
I'm ok with merging

**Comment by taldcroft** [comment]
Woohoo!  Now on to the docs PR?
